httpcache.go

```diff
@@ -498,6 +506,44 @@ func headerAllCommaSepValues(headers http.Header, name string) []string {
 	return vals
 }
 
+type cachedReadCloser struct {
+	resp     *http.Response
+	origBody io.ReadCloser
+	buffer   *bytes.Buffer
+	cache    Cache
+	key      string
+}
+
+func newCachedReadCloser(resp *http.Response, cache Cache, key string) io.ReadCloser {
+	return &cachedReadCloser{
+		resp:     resp,
+		origBody: resp.Body,
+		buffer:   &bytes.Buffer{},
+		cache:    cache,
+		key:      key,
+	}
+}
+
+func (r *cachedReadCloser) Read(p []byte) (n int, err error) {
+	n, err = io.TeeReader(r.origBody, r.buffer).Read(p)
```

Comment:
This is odd. Why create a new `io.TeeReader` on each call to `cachedReadCloser.Read`?

Can you use `io.TeeReader` in `newCachedReadCloser` so it's only called once, returns a reader, and this `Read` method can just use it instead of creating a new reader each time?

Also, the implementation of `io.TeeReader` is [trivial](https://github.com/golang/go/blob/c522b2bec9e6186a8efb7ddd221295390638a18e/src/io/io.go#L525-L531). Maybe it's better and more readable to just embed its functionality into `cachedReadCloser` directly?